### PR TITLE
In str.format, check the security for keys and items that are accessed. [2.13]

### PR DESCRIPTION
**Pull request by mauritsvanrees**
Part of PloneHotfix20171128.

I have some more tests that I could add, but they required PageTemplates, so it is better if I add those in Zope. That is how I did it for the similar January hotfix too, which fixed attribute access in the same way. It may be easier to add those later.

### CODE REVIEW

File: src/AccessControl/tests/test_formatter.py [hunk 1]
```diff
@@ -36,14 +36,46 @@ def test_positional_argument_regression(self):
...
+                          SafeFormatter(u'{0[__private__]}').safe_format,
+                          foo)
+
+    def test_prevents_bad_string_formatting_key(self):
+        from AccessControl.safe_formatter import SafeFormatter
+        from persistent.list import PersistentList
+        foo = PersistentList(['__private__'])
+        self.assertRaises(Unauthorized,
+                          SafeFormatter('{0[0]}').safe_format,
+                          foo)
```
Comment (tseaver) [inline src/AccessControl/tests/test_formatter.py]: Why would we want to protect this string list item?  AFAIK, "values" (rather than names / keys) don't get such protection under our security model.

Comment (mauritsvanrees) [reply src/AccessControl/tests/test_formatter.py]: Say you get a list of items from a folderish content type, with `contentValues()` or maybe a custom method. If you use that list in a Zope2 page template or a Python skin Script, it seems good that AccessControl checks if you are allowed to access an item from that list.

Ah, I started wondering what happens if I try to circumvent this patch by simply getting a list of items, and asking for a private item by index, without bothering to use `str.format`. I see that, also without this patch, checks are done.  I tried this code in a Script Python in a Plone Site:

```
print context.contentValues()[index]
return printed
```

When I execute this script as anonymous user and pass an index that gives me a public page, I can see it just fine. (Well, `<Document at public>` here.) When instead I pass an index that would give me a private page, I get an Unauthorized. When I try that as admin user, I see it fine again.

I can partially get around that by doing a slice, because `guarded_getitem` does not check them: `context.contentValues()[4:]` will give me something like `[<Document at /Plone/private>]`. But that is still only a list, so when I want to actually do something with it, like accessing the Title, I need to get the item, which will properly call `guarded_getitem` again, which will block me.

So it looks like values are checked after all, at least in this example case.

Note that `guarded_getitem` does no checks when both the container and the contained items are simple Python objects, which is why I had to test this with a `PersistentList` instead of a standard Python list.

**Comment by tseaver** [comment]
I still think that excluding string value list items is bogus:  strings are "rocks," and have no reason to be protected.  Checking that an instance-with-private-permissions is excluded is perfectly ok.

**Comment by mauritsvanrees** [comment]
Okay, I have added more extensive tests with non-basic Python types. The most extensive one is `test_prevents_bad_string_formatting_key` as it tests `str.format` in three cases :

1. accessing a basic Python type in a basic Python list: this is allowed
2. accessing a basic Python type in a non-basic PersistentList: this is *not* allowed, unless explicitly allowed by setting `__allow_access_to_unprotected_subobjects__` on the PersistentList.
3. accessing a non-basic persistent Item in a non-basic persistent Folder: the `__roles__` on the Item are checked, and whatever other checks AccessControl does.

Number two in that list might be okay to allow. But the point here is: `str.format` should use the same rules that `guarded_getitem` uses. And `guarded_getitem` does not allow access in the case of number 2. See [`ZopeGuards.py`](https://github.com/zopefoundation/AccessControl/blob/2.13.15/src/AccessControl/ZopeGuards.py#L84-L86): it only allows unchecked access if both the list and the item are simple types.

To summarise for clarity: the only thing this PR changes, is that using `str.format` or `unicode.format` in restricted Python code now explicitly calls `guarded_getitem`.

**Comment by mauritsvanrees** [comment]
@tseaver Are you okay with this, now that I added tests and further explanation above?

File: src/AccessControl/tests/test_formatter.py [hunk 2]
```diff
@@ -36,14 +92,129 @@ def test_positional_argument_regression(self):
...
+        from AccessControl.safe_formatter import SafeFormatter
+        from AccessControl.ZopeGuards import guarded_getitem
+        from persistent.list import PersistentList
+        # Accessing basic Python types in a basic Python list is fine.
+        foo = list(['__private__'])
+        self.assertEqual(SafeFormatter('{0[0]}').safe_format(foo),
+                         '__private__')
+        self.assertEqual(guarded_getitem(foo, 0), '__private__')
+        # For basic Python types in a non-basic list, we guard the access.
+        foo = PersistentList(foo)
```
Comment (tseaver) [inline src/AccessControl/tests/test_formatter.py]: I would prefer that this test *not* use strings like `'__private__'` in this example:  the point of `__allow_access_to_unprotected_subobjects` is that it denies access *regardless* of the contents for items without explicit protection.

Comment (mauritsvanrees) [reply src/AccessControl/tests/test_formatter.py]: Fair point. I have changed the test.

**Comment by gforcada** [comment]
Friendly ping to whoever can click on the merge button 🙂

**Comment by mauritsvanrees** [comment]
I have rebased and squashed to make it cleaner.
@icemac has approved it in December, and I have addressed the helpful critique from @tseaver.
Note that the real fix was (using guarded_getitem) was already done on master in September, and released in 4.0b1.
Given that, I think I will allow myself to merge this soon, if someone doesn't beat me to it.

**Comment by icemac** [comment]
@mauritsvanrees Feel free to merge it!
